### ISSUE TITLE
BUG: `Literal` type decays to `Unknown | str` within a listcomp

### ISSUE THREAD

**Issue by neutrinoceros** [astral-sh/ty#1796]
### Summary

I made an attempt at finding an existing report for this but I think I miss the vocabulary to describe this problem in a searchable way; so apologies if this is a duplicate.

In short, here's an MRE
```py
# t.py
from typing import Literal

Fruit = Literal["banana", "kiwi"]

def eat_fruit(frt: Fruit) -> int:
    print(f"Hum... {frt}")
    return 1

def process(frt: Fruit, other: Fruit):
    a, b = [eat_fruit(f) for f in [frt, other]]
```
I believe this code is perfectly type-consistent, however `ty check` does emit a diagnostic for it

```
❯ ty check t.py
error[invalid-argument-type]: Argument to function `eat_fruit` is incorrect
  --> t.py:10:23
   |
 9 | def process(frt: Fruit, other: Fruit):
10 |     a, b = [eat_fruit(f) for f in [frt, other]]
   |                       ^ Expected `Literal["banana", "kiwi"]`, found `Unknown | str`
   |
info: Element `str` of this union is not assignable to `Literal["banana", "kiwi"]`
info: Function defined here
 --> t.py:5:5
  |
3 | Fruit = Literal["banana", "kiwi"]
4 |
5 | def eat_fruit(frt: Fruit) -> int:
  |     ^^^^^^^^^ ---------- Parameter declared here
6 |     print(f"Hum... {frt}")
7 |     return 1
  |
info: rule `invalid-argument-type` is enabled by default

Found 1 diagnostic
```

### Version

ty 0.0.1-alpha.32 (84a188116 2025-12-05)

**Comment by AlexWaygood** [comment]
Thank you! I believe this is https://github.com/astral-sh/ty/issues/1602. We don't currently propagate type context (provided here by the parameter annotations `frt: Fruit` and `other: Fruit`) from outer scopes into inner scopes. (An inner scope is created implicitly by the list comprehension.) In order to infer the `[frt, other]` list inside the list comprehension as `list[Fruit]` rather than `list[str | Unknown]`, that's what we need to do.

**Comment by neutrinoceros** [comment]
Excellent. Will follow the discussion there. Thanks a lot !
